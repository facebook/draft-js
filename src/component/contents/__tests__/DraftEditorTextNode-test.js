/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 * @format
 */

'use strict';

jest.disableAutomock().mock('UserAgent');

const TEST_A = 'Hello';
const TEST_B = ' World!';

const DraftEditorTextNode = require('DraftEditorTextNode.react');
const React = require('React');
const ReactDOM = require('ReactDOM');
const UserAgent = require('UserAgent');

let container;

beforeEach(() => {
  jest.resetModules();
  container = document.createElement('div');
});

const renderIntoContainer = element => {
  return ReactDOM.render(element, container);
};

const initializeAsIE = () => {
  UserAgent.isBrowser.mockImplementation(() => true);
};

const initializeAsNonIE = () => {
  UserAgent.isBrowser.mockImplementation(() => false);
};

test('must initialize correctly with an empty string, non-IE', () => {
  initializeAsNonIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{''}</DraftEditorTextNode>,
  );
  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must initialize correctly with an empty string, IE', () => {
  initializeAsIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{''}</DraftEditorTextNode>,
  );
  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must initialize correctly with a string, non-IE', () => {
  initializeAsNonIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>,
  );
  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must initialize correctly with a string, IE', () => {
  initializeAsIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>,
  );
  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must update from empty to non-empty, non-IE', () => {
  initializeAsNonIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{''}</DraftEditorTextNode>,
  );

  renderIntoContainer(<DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>);

  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must update from empty to non-empty, IE', () => {
  initializeAsIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{''}</DraftEditorTextNode>,
  );

  renderIntoContainer(<DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>);

  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must update from non-empty to non-empty, non-IE', () => {
  initializeAsIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>,
  );

  renderIntoContainer(
    <DraftEditorTextNode>{TEST_A + TEST_B}</DraftEditorTextNode>,
  );

  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();

  renderIntoContainer(<DraftEditorTextNode>{TEST_B}</DraftEditorTextNode>);

  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must skip updates if text already matches DOM, non-IE', () => {
  initializeAsNonIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>,
  );

  spyOn(stub, 'render').and.callThrough();

  renderIntoContainer(<DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>);

  expect(stub.render.calls.count()).toMatchSnapshot();

  // Sanity check that updating is performed when appropriate.
  renderIntoContainer(<DraftEditorTextNode>{TEST_B}</DraftEditorTextNode>);

  expect(stub.render.calls.count()).toMatchSnapshot();
});

test('must skip updates if text already matches DOM, IE', () => {
  initializeAsIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>,
  );

  spyOn(stub, 'render').and.callThrough();

  renderIntoContainer(<DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>);

  expect(stub.render.calls.count()).toMatchSnapshot();

  // Sanity check that updating is performed when appropriate.
  renderIntoContainer(<DraftEditorTextNode>{TEST_B}</DraftEditorTextNode>);

  expect(stub.render.calls.count()).toMatchSnapshot();
});

test('must update from non-empty to empty, non-IE', () => {
  initializeAsNonIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>,
  );

  renderIntoContainer(<DraftEditorTextNode>{''}</DraftEditorTextNode>);

  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must update from non-empty to empty, IE', () => {
  initializeAsIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>,
  );

  renderIntoContainer(<DraftEditorTextNode>{''}</DraftEditorTextNode>);

  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must render properly into a parent DOM node', () => {
  initializeAsNonIE();

  const stub = renderIntoContainer(
    <div>
      <DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>
    </div>,
  );

  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});

test('must force unchanged text back into the DOM', () => {
  initializeAsNonIE();
  const stub = renderIntoContainer(
    <DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>,
  );

  ReactDOM.findDOMNode(stub).textContent = TEST_B;

  renderIntoContainer(<DraftEditorTextNode>{TEST_A}</DraftEditorTextNode>);

  expect(ReactDOM.findDOMNode(stub)).toMatchSnapshot();
});
