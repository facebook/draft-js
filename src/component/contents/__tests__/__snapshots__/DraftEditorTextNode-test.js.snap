// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must force unchanged text back into the DOM 1`] = `
<span
  data-reactroot=""
  data-text="true"
>
  Hello
</span>
`;

exports[`must initialize correctly with a string, IE 1`] = `
<span
  data-reactroot=""
  data-text="true"
>
  Hello
</span>
`;

exports[`must initialize correctly with a string, non-IE 1`] = `
<span
  data-reactroot=""
  data-text="true"
>
  Hello
</span>
`;

exports[`must initialize correctly with an empty string, IE 1`] = `
<span
  data-reactroot=""
  data-text="true"
>
  

</span>
`;

exports[`must initialize correctly with an empty string, non-IE 1`] = `
<br
  data-reactroot=""
  data-text="true"
/>
`;

exports[`must render properly into a parent DOM node 1`] = `
<div
  data-reactroot=""
>
  <span
    data-text="true"
  >
    Hello
  </span>
</div>
`;

exports[`must skip updates if text already matches DOM, IE 1`] = `0`;

exports[`must skip updates if text already matches DOM, IE 2`] = `1`;

exports[`must skip updates if text already matches DOM, non-IE 1`] = `0`;

exports[`must skip updates if text already matches DOM, non-IE 2`] = `1`;

exports[`must update from empty to non-empty, IE 1`] = `
<span
  data-reactroot=""
  data-text="true"
>
  Hello
</span>
`;

exports[`must update from empty to non-empty, non-IE 1`] = `
<span
  data-reactroot=""
  data-text="true"
>
  Hello
</span>
`;

exports[`must update from non-empty to empty, IE 1`] = `
<span
  data-reactroot=""
  data-text="true"
>
  

</span>
`;

exports[`must update from non-empty to empty, non-IE 1`] = `
<br
  data-reactroot=""
  data-text="true"
/>
`;

exports[`must update from non-empty to non-empty, non-IE 1`] = `
<span
  data-reactroot=""
  data-text="true"
>
  Hello World!
</span>
`;

exports[`must update from non-empty to non-empty, non-IE 2`] = `
<span
  data-reactroot=""
  data-text="true"
>
   World!
</span>
`;
